collector.go

```diff
@@ -50,10 +56,64 @@ func NewCollector(config CollectorConfig) (*Collector, error) {
 }
 
 func (c *Collector) Bundles() []Bundle {
-	return c.bundles
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	return CopyBundles(c.bundles)
+}
+
+type CollectorEndpointResponse struct {
+	VersionBundles []Bundle `json:"version_bundles"`
 }
 
-// TODO
-func (c *Collector) Collect() error {
+func (c *Collector) Collect(ctx context.Context) error {
+	var responses [][]byte
+	{
+		var g errgroup.Group
+		responses = make([][]byte, len(c.endpoints))
+
+		for i, e := range c.endpoints {
+			i, e := i, e
```

Comment:
?

Reply:
See the godoc example: https://godoc.org/golang.org/x/sync/errgroup#example-Group--Parallel. It refers to https://golang.org/doc/faq#closures_and_goroutines. For loop values work with pointers under the hood. To be able to reliably work with values within goroutines you need to pin the value to your goroutine scope. Otherwise most goroutines would only see the value of the last for loop iteration.

Reply:
You can make it more obvious (IMO) by passing them as the anonymous func arguments.

Reply:
Show me how. This does not work with errgroup because you just execute `func () error`. You cannot inject anything there. This code here applies to the godoc examples.

Reply:
> This does not work with errgroup because you just execute func () error.

You're right I haven't realise that.